README.md
AWS Transcribe Streaming client for producing real-time transcriptions for your media content using HTTP/2.

## Getting started

Add necessary dependencies to `pubspec.yaml`:
```yaml
dependencies:
  aws_common: ^0.6.0
  aws_transcribe_streaming: ^0.1.0
```

Obtain a pair of Access/Secret keys for the AWS IAM user with `transcribe:StartStreamTranscription` permission.
See details in AWS documentation: [Transcribing streaming audio](https://docs.aws.amazon.com/transcribe/latest/dg/streaming.html)

It is recommended to use [Temporary security credentials](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp.html) with session token obtained from the backend just before starting the transcribing process.

See also [best practices](https://docs.aws.amazon.com/transcribe/latest/dg/streaming.html#best-practices) to improve streaming transcription efficiency.

## Usage

1. Create a new transcribe streaming client:
```dart
import 'package:aws_common/aws_common.dart';
import 'package:aws_transcribe_streaming/aws_transcribe_streaming.dart';

final transcribeStreamingClient = TranscribeStreamingClient(
  region: 'eu-central-1',
  credentialsProvider: StaticCredentialsProvider(AWSCredentials(
    'ASIAIOEXAMPLEEXAMPLE',                       // accessKeyId
    'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY',   // secretAccessKey
    'AQoDYXdzEJr...',                             // sessionToken
    DateTime.now().add(const Duration(hours: 1)), // expiration
  )),
);
```

2. Start a stream transcription:
```dart
final (response, audioStreamSink, transcriptEventStream) =
    await transcribeStreamingClient.startStreamTranscription(
  const StartStreamTranscriptionRequest(
    languageCode: LanguageCode.enUs,
    mediaSampleRateHertz: 48000,
    mediaEncoding: MediaEncoding.pcm,
  ),
);
```

3. Subscribe to a raw TranscriptEvent stream:
```dart
final transcriptSubscription = transcriptEventStream
    .listen((TranscriptEvent event) => print(event));
```
or use a custom strategy to decode TranscriptEvents and build the realtime transcription:
```dart
final transcriptSubscription = transcriptEventStream
    .transform(const TranscriptEventStreamDecoder(PlainTextTranscriptionStrategy()))
    .listen((String message) => print(message));
```

4. Start sending audio data to the audio stream sink:
```dart
// Raw audio data from the microphone in PCM signed 16-bit little-endian audio format
Stream<Uint8List> audioStream;

final audioStreamSubscription = audioStream.listen(
  audioStreamSink.add,
  onDone: audioStreamSink.close,
);
```


CHANGELOG.md
## 0.1.0

* First public release.


lib/src/event_stream/header_codec.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:uuid/uuid.dart';

import 'exceptions.dart';
import 'header.dart';

/// Possible types of [EventStreamHeader].
enum EventStreamHeaderType {
  boolTrue,
  boolFalse,
  byte,
  short,
  integer,
  long,
  byteArray,
  string,
  timestamp,
  uuid,
}

/// Codec for encoding and decoding [EventStreamHeader].
final class EventStreamHeaderCodec {
  /// Decodes [Uint8List] into [EventStreamHeader].
  static List<EventStreamHeader> decode(Uint8List bytes) {
    final out = <EventStreamHeader>[];
    int position = 0;
    final headers = ByteData.view(bytes.buffer);

    while (position < headers.lengthInBytes) {
      final nameLength = headers.getUint8(position++);
      final name = utf8.decode(bytes.sublist(position, position + nameLength));
      position += nameLength;

      final type = EventStreamHeaderType.values[headers.getUint8(position++)];
      late EventStreamHeader header;

      switch (type) {
        case EventStreamHeaderType.boolTrue:
          header = EventStreamBoolHeader(name, true);
          break;
        case EventStreamHeaderType.boolFalse:
          header = EventStreamBoolHeader(name, false);
          break;
        case EventStreamHeaderType.byte:
          header = EventStreamByteHeader(name, headers.getInt8(position));
          position++;
          break;
        case EventStreamHeaderType.short:
          header = EventStreamShortHeader(
            name,
            headers.getInt16(position, Endian.big),
          );
          position += 2;
          break;
        case EventStreamHeaderType.integer:
          header = EventStreamIntegerHeader(
            name,
            headers.getInt32(position, Endian.big),
          );
          position += 4;
          break;
        case EventStreamHeaderType.long:
          header = EventStreamLongHeader(
            name,
            headers.getInt64(position, Endian.big),
          );
          position += 8;
          break;
        case EventStreamHeaderType.byteArray:
          final binaryLength = headers.getUint16(position, Endian.big);
          position += 2;
          header = EventStreamByteArrayHeader(
            name,
            bytes.sublist(position, position + binaryLength),
          );
          position += binaryLength;
          break;
        case EventStreamHeaderType.string:
          final stringLength = headers.getUint16(position, Endian.big);
          position += 2;
          header = EventStreamStringHeader(
            name,
            utf8.decode(bytes.sublist(position, position + stringLength)),
          );
          position += stringLength;
          break;
        case EventStreamHeaderType.timestamp:
          header = EventStreamTimestampHeader(
            name,
            DateTime.fromMillisecondsSinceEpoch(
              headers.getInt64(position, Endian.big),
            ),
          );
          position += 8;
          break;
        case EventStreamHeaderType.uuid:
          header = EventStreamUuidHeader(
            name,
            Uuid.unparse(bytes.sublist(position, position + 16)),
          );
          position += 16;
          break;
        default:
          throw const EventStreamHeaderDecodeException(
            'Unrecognized header type',
          );
      }

      out.add(header);
    }

    return out;
  }

  /// Encodes a list of [EventStreamHeader]s into [Uint8List].
  static Uint8List encodeHeaders(List<EventStreamHeader> headers) {
    final bytes = <int>[];

    for (final header in headers) {
      bytes.addAll(EventStreamHeaderCodec.encodeHeader(header));
    }

    return Uint8List.fromList(bytes);
  }

  /// Encodes a single [EventStreamHeader] into [Uint8List].
  static Uint8List encodeHeader(EventStreamHeader header) {
    final bytes = <int>[];

    final nameBytes = utf8.encode(header.name);
    bytes.add(nameBytes.length);
    bytes.addAll(nameBytes);

    switch (header) {
      case EventStreamBoolHeader(:var value):
        bytes.add(
          value
              ? EventStreamHeaderType.boolTrue.index
              : EventStreamHeaderType.boolFalse.index,
        );
        break;
      case EventStreamByteHeader(:var value):
        bytes.add(EventStreamHeaderType.byte.index);
        bytes.add(value);
        break;
      case EventStreamShortHeader(:var value):
        bytes.add(EventStreamHeaderType.short.index);
        bytes.addAll(_encodeInt16(value));
        break;
      case EventStreamIntegerHeader(:var value):
        bytes.add(EventStreamHeaderType.integer.index);
        bytes.addAll(_encodeInt32(value));
        break;
      case EventStreamLongHeader(:var value):
        bytes.add(EventStreamHeaderType.long.index);
        bytes.addAll(_encodeInt64(value));
        break;
      case EventStreamByteArrayHeader(:var value):
        bytes.add(EventStreamHeaderType.byteArray.index);
        bytes.addAll(_encodeUint16(value.lengthInBytes));
        bytes.addAll(value);
        break;
      case EventStreamStringHeader(:var value):
        bytes.add(EventStreamHeaderType.string.index);
        final valueBytes = utf8.encode(value);
        bytes.addAll(_encodeUint16(valueBytes.length));
        bytes.addAll(valueBytes);
        break;
      case EventStreamTimestampHeader(:var value):
        bytes.add(EventStreamHeaderType.timestamp.index);
        bytes.addAll(_encodeInt64(value.millisecondsSinceEpoch));
        break;
      case EventStreamUuidHeader(:var value):
        bytes.add(EventStreamHeaderType.uuid.index);
        bytes.addAll(Uuid.parse(value));
        break;
    }

    return Uint8List.fromList(bytes);
  }

  static Uint8List _encodeUint16(int value) {
    final bytes = Uint8List(2);
    ByteData.view(bytes.buffer).setUint16(0, value, Endian.big);
    return bytes;
  }

  static Uint8List _encodeInt16(int value) {
    final bytes = Uint8List(2);
    ByteData.view(bytes.buffer).setInt16(0, value, Endian.big);
    return bytes;
  }

  static Uint8List _encodeInt32(int value) {
    final bytes = Uint8List(4);
    ByteData.view(bytes.buffer).setInt32(0, value, Endian.big);
    return bytes;
  }

  static Uint8List _encodeInt64(int value) {
    final bytes = Uint8List(8);
    ByteData.view(bytes.buffer).setInt64(0, value, Endian.big);
    return bytes;
  }
}


example/main.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:aws_common/aws_common.dart';
import 'package:aws_transcribe_streaming/aws_transcribe_streaming.dart';

void main() async {
  // Create a client.
  final transcribeStreamingClient = TranscribeStreamingClient(
    region: 'eu-central-1',
    // Provide credentials with `transcribe:StartStreamTranscription` permission
    credentialsProvider: StaticCredentialsProvider(AWSCredentials(
      'ASIAIOEXAMPLEEXAMPLE', // accessKeyId
      'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY', // secretAccessKey
      'AQoDYXdzEJr...', // sessionToken
      DateTime.now().add(const Duration(hours: 1)), // expiration
    )),
  );

  late final StartStreamTranscriptionResponse response;
  late final Stream<TranscriptEvent> transcriptEventStream;
  late final StreamSink<Uint8List> audioStreamSink;
  StreamSubscription<Uint8List>? audioStreamSubscription;

  try {
    // Start a stream transcription.
    (response, audioStreamSink, transcriptEventStream) =
        await transcribeStreamingClient.startStreamTranscription(
      const StartStreamTranscriptionRequest(
        languageCode: LanguageCode.enUs,
        mediaSampleRateHertz: 48000,
        mediaEncoding: MediaEncoding.pcm,
      ),
    );
  } on TranscribeStreamingException catch (e) {
    print('Error starting transcription: $e');
    return;
  }

  print('Session ID: ${response.sessionId}');

  final transcriptionCompleter = Completer<void>();

  // Listen to transcript events.
  // final transcriptSubscription =
  //     transcriptEventStream.listen((TranscriptEvent event) => print(event));
  // or use a custom strategy to decode transcript events.
  transcriptEventStream
      .transform(
          const TranscriptEventStreamDecoder(PlainTextTranscriptionStrategy()))
      .listen(
    (String message) {
      print('Transcription: $message');
    },
    onError: (Object error, StackTrace stackTrace) async {
      print('Transcription error: $error');
      await audioStreamSubscription?.cancel();
      await audioStreamSink.close();
    },
    onDone: () async {
      print('Transcription done');
      await audioStreamSubscription?.cancel();
      transcriptionCompleter.complete();
    },
  );

  // Instead use a real stream of audio data from the microphone
  // in PCM signed 16-bit little-endian audio format with 48kHz sample rate.
  final audioStream = Stream<Uint8List>.periodic(
      const Duration(milliseconds: 200), (count) => Uint8List(19200)).take(25);

  // Send audio data to the audio stream sink.
  audioStreamSubscription = audioStream.listen(
    audioStreamSink.add,
    onDone: audioStreamSink.close,
  );

  await transcriptionCompleter.future;

  print('Finished');
}
